Fix transitive derivation for ConfiguredEncoder

MartinHH opened the pull request:
Fixes that derivation would not use existing givens if those depend on a given that needs to be derived (e.g. not using `encodeOption[A]` if the required `Encoder[A]` needs to be derived).

(Problem still prevails for all other derivable typeclasses.)

Fixes #2212

MartinHH committed:
Fix transitive derivation for ConfiguredEncoder

Fixes that derivation would not use existing givens if those depend on
a given that needs to be derived (e.g. not using `encodeOption[A]` if
the required `Encoder[A]` needs to be derived).

(Problem still prevails for all other derivable typeclasses.)

Fixes #2212

MartinHH commented:
If this solution approach gets approved, I would be willing to roll this out for all other typeclasses that are affected by the same problem.

MartinHH commented on modules/core/shared/src/main/scala-3/io/circe/derivation/ConfiguredEncoder.scala:
```diff
@@ -54,8 +55,17 @@ trait ConfiguredEncoder[A](using conf: Configuration) extends Encoder.AsObject[A
+
   inline final def derived[A](using conf: Configuration)(using inline mirror: Mirror.Of[A]): ConfiguredEncoder[A] =
     new ConfiguredEncoder[A] with SumOrProduct:
+      import GivenDerivation.deriveEncoder
```
Side note: theoretically, this would allow to change the implementation of `io.circe.derivation.summonEncoder` from...
```scala
private[circe] inline final def summonEncoder[A](using Configuration): Encoder[A] =
  summonFrom {
    case encodeA: Encoder[A] => encodeA
    case _: Mirror.Of[A]     => ConfiguredEncoder.derived[A]
  }
```
...to...
```scala
private[circe] inline final def summonEncoder[A](using Configuration): Encoder[A] =
  summonInline[Encoder[A]]
```
...because derivation via `Mirror.Of[A]` is now available implictly so `summonInline[Encoder[A]]` will find it. (I tried that and all tests ran successfully.)

However: that code is also accessible via (deprecated) public API (see `io.circe.Derivation`) so I guess it's better to leave it as it is.

zarthross commented on modules/tests/shared/src/test/scala-3/io/circe/ConfiguredDerivesSuite.scala:
```diff
@@ -504,3 +504,21 @@ class ConfiguredDerivesSuite extends CirceMunitSuite:
+
+    case class Bar(barValue: Int)
+    case class Foo(fooValue: Option[Bar]) derives ConfiguredEncoder
+
```
Can you add a codec law here for `Foo` ?
```suggestion
checkAll("Codec[Foo]", CodecTests[Foo].codec)
```

MartinHH replied:
So I rolled out my solution approach onto the other type classes so that I can check codec laws (see new commit), but it turns out that "codec.encoder serializability" and "codec.decoder serializability" fail.

I will look into that when I find time, but if anyone has experience with that, any pointers would be welcome. (Been years since I last had to mess with java object serialization...)

MartinHH replied:
> So I rolled out my solution approach onto the other type classes so that I can check codec laws (see new commit), but it turns out that "codec.encoder serializability" and "codec.decoder serializability" fail.
> 
> I will look into that when I find time, but if anyone has experience with that, any pointers would be welcome. (Been years since I last had to mess with java object serialization...)

Nevermind, I fixed it.

MartinHH committed:
Fix transitive derivation for all Codecs

Fixes that derivation would not use existing givens if those depend on
a given that needs to be derived (e.g. not using `encodeOption[A]` if
the required `Encoder[A]` needs to be derived).

(Still has tests failing due to NotSerializableException.)

Fixes #2212

MartinHH committed:
Fix NotSerializableException in tests

Fixes #2212
